File: internal/handlers/contact_handler.go
package handlers

import (
	"context"
	"net/http"
	"time"

	"chat-backend/internal/database"
	"chat-backend/internal/models"
	"chat-backend/internal/utils"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ContactHandler struct {
	db *database.Database
}

func NewContactHandler(db *database.Database) *ContactHandler {
	return &ContactHandler{db: db}
}

type ScanQRRequest struct {
	QRData string `json:"qr_data" binding:"required"`
}

type AddContactRequest struct {
	UserID      string `json:"user_id"`      // optional when phone_number+display_name given
	PhoneNumber string `json:"phone_number"` // add by number
	DisplayName string `json:"display_name"` // name for the contact
}

func (h *ContactHandler) GetContacts(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	cursor, err := h.db.MongoDB.Collection("contacts").Find(
		context.Background(),
		bson.M{"user_id": userIDObj},
	)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch contacts"})
		return
	}
	defer cursor.Close(context.Background())

	var contacts []models.Contact
	if err := cursor.All(context.Background(), &contacts); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to decode contacts"})
		return
	}

	// Fetch user details for each contact (or use phone_number/display_name for external contacts)
	var contactDetails []map[string]interface{}
	for _, contact := range contacts {
		if contact.ContactID != nil {
			var user models.User
			err := h.db.MongoDB.Collection("users").FindOne(
				context.Background(),
				bson.M{"_id": *contact.ContactID},
			).Decode(&user)
			if err == nil {
				contactDetails = append(contactDetails, map[string]interface{}{
					"contact": contact,
					"user":    user,
				})
			}
		} else {
			// External contact (added by phone + name only)
			contactDetails = append(contactDetails, map[string]interface{}{
				"contact": contact,
				"user": map[string]interface{}{
					"id":           nil,
					"phone_number": contact.PhoneNumber,
					"username":     contact.DisplayName,
					"display_name": contact.DisplayName,
				},
			})
		}
	}

	c.JSON(http.StatusOK, contactDetails)
}

// AddContact adds a contact by user_id OR by phone_number + display_name.
func (h *ContactHandler) AddContact(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req AddContactRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Add by phone_number + display_name
	if req.PhoneNumber != "" {
		phone := normalizePhone(req.PhoneNumber)
		if phone == "" {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid phone number"})
			return
		}
		// Check if already in contacts (by phone or by user)
		var existing models.Contact
		err := h.db.MongoDB.Collection("contacts").FindOne(
			context.Background(),
			bson.M{"user_id": userIDObj, "phone_number": phone},
		).Decode(&existing)
		if err == nil {
			c.JSON(http.StatusConflict, gin.H{"error": "Contact already exists"})
			return
		}
		var contactUser models.User
		err = h.db.MongoDB.Collection("users").FindOne(
			context.Background(),
			bson.M{"phone_number": phone},
		).Decode(&contactUser)
		displayName := req.DisplayName
		if displayName == "" {
			displayName = phone
		}
		if err != nil {
			// User not registered: add as external contact (no ContactID)
			contact := models.Contact{
				ID:          primitive.NewObjectID(),
				UserID:      userIDObj,
				ContactID:   nil,
				PhoneNumber: phone,
				DisplayName: displayName,
				IsAnonymous: false,
				CreatedAt:   time.Now(),
			}
			if _, err := h.db.MongoDB.Collection("contacts").InsertOne(context.Background(), contact); err != nil {
				c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to add contact"})
				return
			}
			c.JSON(http.StatusCreated, contact)
			return
		}
		// User registered: add with ContactID
		if contactUser.ID == userIDObj {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Cannot add yourself as contact"})
			return
		}
		err = h.db.MongoDB.Collection("contacts").FindOne(
			context.Background(),
			bson.M{"user_id": userIDObj, "contact_id": contactUser.ID},
		).Decode(&existing)
		if err == nil {
			c.JSON(http.StatusConflict, gin.H{"error": "Contact already exists"})
			return
		}
		contact := models.Contact{
			ID:          primitive.NewObjectID(),
			UserID:      userIDObj,
			ContactID:   &contactUser.ID,
			ContactQR:   contactUser.QRCode,
			DisplayName: displayName,
			IsAnonymous: false,
			CreatedAt:   time.Now(),
		}
		if _, err := h.db.MongoDB.Collection("contacts").InsertOne(context.Background(), contact); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to add contact"})
			return
		}
		reverseContact := models.Contact{
			ID:          primitive.NewObjectID(),
			UserID:      contactUser.ID,
			ContactID:   &userIDObj,
			IsAnonymous: false,
			CreatedAt:   time.Now(),
		}
		h.db.MongoDB.Collection("contacts").InsertOne(context.Background(), reverseContact)
		c.JSON(http.StatusCreated, contact)
		return
	}

	// Add by user_id (existing flow)
	if req.UserID == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Provide user_id or phone_number + display_name"})
		return
	}
	contactUserID, err := primitive.ObjectIDFromHex(req.UserID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID"})
		return
	}
	if contactUserID == userIDObj {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Cannot add yourself as contact"})
		return
	}
	var contactUser models.User
	if err := h.db.MongoDB.Collection("users").FindOne(
		context.Background(),
		bson.M{"_id": contactUserID},
	).Decode(&contactUser); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}
	var existingContact models.Contact
	err = h.db.MongoDB.Collection("contacts").FindOne(
		context.Background(),
		bson.M{"user_id": userIDObj, "contact_id": contactUserID},
	).Decode(&existingContact)
	if err == nil {
		c.JSON(http.StatusConflict, gin.H{"error": "Contact already exists"})
		return
	}
	contact := models.Contact{
		ID:          primitive.NewObjectID(),
		UserID:      userIDObj,
		ContactID:   &contactUserID,
		ContactQR:   contactUser.QRCode,
		IsAnonymous: false,
		CreatedAt:   time.Now(),
	}
	if _, err := h.db.MongoDB.Collection("contacts").InsertOne(context.Background(), contact); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to add contact"})
		return
	}
	reverseContact := models.Contact{
		ID:          primitive.NewObjectID(),
		UserID:      contactUserID,
		ContactID:   &userIDObj,
		IsAnonymous: false,
		CreatedAt:   time.Now(),
	}
	h.db.MongoDB.Collection("contacts").InsertOne(context.Background(), reverseContact)
	c.JSON(http.StatusCreated, contact)
}

func normalizePhone(s string) string {
	var out []rune
	for _, r := range s {
		if r >= '0' && r <= '9' || r == '+' {
			out = append(out, r)
		}
	}
	return string(out)
}

func (h *ContactHandler) ScanQRCode(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	var req ScanQRRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Parse QR code
	contactUserIDStr, err := utils.ParseQRCode(req.QRData)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid QR code"})
		return
	}

	contactUserID, err := primitive.ObjectIDFromHex(contactUserIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid user ID in QR code"})
		return
	}

	// Check if contact already exists
	var existingContact models.Contact
	err = h.db.MongoDB.Collection("contacts").FindOne(
		context.Background(),
		bson.M{
			"user_id":    userIDObj,
			"contact_id": contactUserID,
		},
	).Decode(&existingContact)

	if err == nil {
		c.JSON(http.StatusConflict, gin.H{"error": "Contact already exists"})
		return
	}

	contactUserIDPtr := &contactUserID
	contact := models.Contact{
		ID:          primitive.NewObjectID(),
		UserID:      userIDObj,
		ContactID:   contactUserIDPtr,
		ContactQR:   req.QRData,
		IsAnonymous: false,
		CreatedAt:   time.Now(),
	}

	_, err = h.db.MongoDB.Collection("contacts").InsertOne(context.Background(), contact)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to add contact"})
		return
	}

	userIDObjPtr := &userIDObj
	reverseContact := models.Contact{
		ID:          primitive.NewObjectID(),
		UserID:      contactUserID,
		ContactID:   userIDObjPtr,
		ContactQR:   "",
		IsAnonymous: false,
		CreatedAt:   time.Now(),
	}
	h.db.MongoDB.Collection("contacts").InsertOne(context.Background(), reverseContact)

	c.JSON(http.StatusCreated, contact)
}

func (h *ContactHandler) DeleteContact(c *gin.Context) {
	userID, _ := c.Get("user_id")
	userIDObj := userID.(primitive.ObjectID)

	contactIDStr := c.Param("contact_id")
	contactID, err := primitive.ObjectIDFromHex(contactIDStr)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid contact ID"})
		return
	}

	// Try delete by contact record _id first (for external contacts or when frontend sends contact.id)
	result, err := h.db.MongoDB.Collection("contacts").DeleteOne(
		context.Background(),
		bson.M{"_id": contactID, "user_id": userIDObj},
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete contact"})
		return
	}
	if result.DeletedCount == 0 {
		// Else treat as "other user" id (contact_id)
		result, err = h.db.MongoDB.Collection("contacts").DeleteOne(
			context.Background(),
			bson.M{"user_id": userIDObj, "contact_id": contactID},
		)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete contact"})
			return
		}
	}
	if result.DeletedCount == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Contact not found"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Contact deleted successfully"})
}

File: src/handlers/contactHandler.js
'use strict';

const { ObjectId } = require('mongodb');
const { getDB } = require('../database');

// ─── Handlers ─────────────────────────────────────────────────────────────────

async function getContacts(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;

    const contacts = await db.collection('contacts').find({ owner_id: userId }).toArray();

    const result = [];
    for (const contact of contacts) {
      let userInfo = null;
      if (contact.user_id) {
        try {
          const user = await db.collection('users').findOne(
            { _id: new ObjectId(contact.user_id) },
            { projection: { _id: 1, username: 1, phone_number: 1, avatar: 1 } }
          );
          if (user) {
            userInfo = {
              id: user._id.toString(),
              username: user.username,
              phone_number: user.phone_number,
              avatar: user.avatar,
            };
          }
        } catch (_) {
          // ignore
        }
      }

      result.push({
        contact_id: contact._id.toString(),
        user_id: contact.user_id || null,
        display_name: contact.display_name,
        user: userInfo,
        created_at: contact.created_at,
      });
    }

    return res.json(result);
  } catch (err) {
    console.error('getContacts error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function addContact(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const { user_id, phone_number, display_name } = req.body || {};

    const now = new Date();
    let targetUserId = null;
    let resolvedDisplayName = display_name || null;

    if (user_id) {
      // Look up user by ID
      let targetUser;
      try {
        targetUser = await db.collection('users').findOne({ _id: new ObjectId(user_id) });
      } catch (_) {
        return res.status(400).json({ error: 'Invalid user_id' });
      }
      if (!targetUser) return res.status(404).json({ error: 'User not found' });
      targetUserId = targetUser._id.toString();
      if (!resolvedDisplayName) {
        resolvedDisplayName = targetUser.username || targetUser.phone_number;
      }
    } else if (phone_number) {
      // Look up user by phone
      const targetUser = await db.collection('users').findOne({ phone_number });
      if (targetUser) {
        targetUserId = targetUser._id.toString();
        if (!resolvedDisplayName) resolvedDisplayName = targetUser.username || phone_number;
      } else {
        resolvedDisplayName = resolvedDisplayName || phone_number;
      }
    } else {
      return res.status(400).json({ error: 'user_id or phone_number is required' });
    }

    // Create contact for current user
    const contactDoc = {
      owner_id: userId,
      user_id: targetUserId,
      display_name: resolvedDisplayName,
      phone_number: phone_number || null,
      created_at: now,
      updated_at: now,
    };
    const insertResult = await db.collection('contacts').insertOne(contactDoc);
    contactDoc._id = insertResult.insertedId;

    // Create reverse contact (bidirectional) if targetUserId exists and is different
    if (targetUserId && targetUserId !== userId) {
      // Find current user info for reverse display name
      let currentUser;
      try {
        currentUser = await db.collection('users').findOne({ _id: new ObjectId(userId) });
      } catch (_) {}

      const reverseDoc = {
        owner_id: targetUserId,
        user_id: userId,
        display_name: currentUser ? (currentUser.username || currentUser.phone_number) : userId,
        phone_number: currentUser ? currentUser.phone_number : null,
        created_at: now,
        updated_at: now,
      };

      // Only insert reverse if it doesn't already exist
      const existing = await db.collection('contacts').findOne({
        owner_id: targetUserId,
        user_id: userId,
      });
      if (!existing) {
        await db.collection('contacts').insertOne(reverseDoc);
      }
    }

    return res.status(201).json({ ...contactDoc, id: insertResult.insertedId.toString() });
  } catch (err) {
    console.error('addContact error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function scanQRCode(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const { qr_data } = req.body || {};

    if (!qr_data) return res.status(400).json({ error: 'qr_data is required' });

    // Expected format: "CHATAPP:userId:uuid"
    const parts = qr_data.split(':');
    if (parts.length < 2 || parts[0] !== 'CHATAPP') {
      return res.status(400).json({ error: 'Invalid QR code format' });
    }

    const targetUserIdStr = parts[1];

    let targetUser;
    try {
      targetUser = await db.collection('users').findOne({ _id: new ObjectId(targetUserIdStr) });
    } catch (_) {
      return res.status(400).json({ error: 'Invalid user ID in QR code' });
    }

    if (!targetUser) return res.status(404).json({ error: 'User not found' });
    if (targetUser._id.toString() === userId) {
      return res.status(400).json({ error: 'Cannot add yourself as contact' });
    }

    const now = new Date();
    const targetUserId = targetUser._id.toString();

    // Upsert contact
    const existing = await db.collection('contacts').findOne({
      owner_id: userId,
      user_id: targetUserId,
    });

    if (!existing) {
      const contactDoc = {
        owner_id: userId,
        user_id: targetUserId,
        display_name: targetUser.username || targetUser.phone_number || targetUserId,
        phone_number: targetUser.phone_number || null,
        created_at: now,
        updated_at: now,
      };
      await db.collection('contacts').insertOne(contactDoc);
    }

    return res.json({
      user: {
        id: targetUser._id.toString(),
        username: targetUser.username,
        phone_number: targetUser.phone_number,
        avatar: targetUser.avatar,
      },
    });
  } catch (err) {
    console.error('scanQRCode error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

async function deleteContact(req, res) {
  try {
    const db = getDB();
    const userId = req.userId;
    const contactIdStr = req.params.contact_id;

    let contactObjId;
    try {
      contactObjId = new ObjectId(contactIdStr);
    } catch (_) {
      return res.status(400).json({ error: 'Invalid contact ID' });
    }

    const deleteResult = await db.collection('contacts').deleteOne({
      _id: contactObjId,
      owner_id: userId,
    });

    if (deleteResult.deletedCount === 0) {
      return res.status(404).json({ error: 'Contact not found' });
    }

    return res.json({ message: 'Contact deleted' });
  } catch (err) {
    console.error('deleteContact error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
}

module.exports = {
  getContacts,
  addContact,
  scanQRCode,
  deleteContact,
};
